server/src/services/InitGoogle.ts

```diff
@@ -92,36 +92,46 @@ async function onTokens(tokens: Credentials) {
       update,
       create: { id: TOKENS_ID, ...update },
     });
-  } else {
+  } else if (existingGoogleTokens) {
     const update = { access_token, expiry_date, is_valid: true };
-    // TODO: Handle the case where the refresh token is not sent *and* the
-    // record doesn't exist. If this happens, we need to redirect them to a
-    // new auth url, but with prompt: 'consent', so that Google will provide
-    // a new refresh token.
     await prisma.google_tokens.update({
       where: { id: TOKENS_ID },
       data: { ...update },
     });
+  } else {
+    // TODO: Handle the case where the refresh token is not sent *and* the
+    // record doesn't exist. If this happens, we need to redirect them to a
+    // new auth url, but with prompt: 'consent', so that Google will provide
+    // a new refresh token.
+    throw new Error('Missing refresh_token');
   }
 }
 
 export async function requestTokens(code: string) {
-  const oauth2Client = createOAuth2Client().on('tokens', onTokens);
-
-  try {
-    await oauth2Client.getToken(code);
-  } catch {
-    throw new Error('Failed to get tokens');
-  }
+  const oauth2Client = createOAuth2Client();
+  await Promise.all([
+    new Promise((resolve, reject) => {
+      oauth2Client.on('tokens', (tokens: Credentials) => {
+        onTokens(tokens).then(resolve).catch(reject);
+      });
+    }),
+    new Promise((resolve, reject) => {
+      oauth2Client
+        .getToken(code)
+        .then(resolve)
+        .catch(() => reject(new Error('Failed to get tokens')));
+    }),
+  ]);
```

Comment:
I see what you mean.  This is tricky.

I'll sleep on it, but what I think is that I'd do this:
```suggestion

  try {
    const tokens = (await oauth2Client.getToken(code)).tokens;
    await onTokens(tokens);
  } catch (e) {
    console.error('Failed to get tokens');
    throw redactSecrets(e);
  }
```

I'm not sure what we should be doing to _other_ `on('tokens')` errors. Ideally we should communicate the issue to the client, somehow, but I'm not sure we want to be messing around with websockets just yet!

Reply:
Thank you, it's much better now.

Because there are situations when lack of refresh_token is okay, I added optional parameter for it.